correct unit cap (cap 12) pin content

Lastdayends opened the pull request:
@stevenmburns 

Although visually there is no via bug in the switched capacitor combination, I did find the bug exists. 

The reason is due to the error in the lef file.
After the M1 port in cap_12 is removed, the via bug should be fixed. (There is an assumption, only M2 pins are provided for unit cap.)

By the way, the location of the via generated by the layout visualization tool is not correct.

Yaguang

Lastdayends commented:
@stevenmburns 

There is a logic bug when writing the internal metal of cap. The off grid M1-M3 in unit cap can not be mapped to any grid point so equivalently always be ignored. That's why the shortage happens. To avoid this, I just block all the M1-M3 of all unit cap area. Here is the result of net vin_ota, which used to have a shortage issue.

![image](https://user-images.githubusercontent.com/43042278/67164213-ad721400-f33d-11e9-9dd1-a90ad4c6f316.png)


If any issues, please let me know.

Yaguang

stevenmburns commented:
can you tell me more about this?
"By the way, the location of the via generated by the layout visualization tool is not correct."
We need to get this right.

Lastdayends commented:
Hi Steve,

As you mentioned 'Still many opens and shorts in wiring up the cap arrays',
are some vias missed? (At the beginning, I saw the same problem, a lot of
vias are missed. After I selected intervias to be shown, the problem is
gone. The vias in cap array are intervias, and needs to be shown to avoid
visually opens or shorts.)

Yaguang


On Mon, Oct 21, 2019 at 9:33 AM Steve Burns <notifications@github.com>
wrote:

> *@stevenmburns* approved this pull request.
>
> Fixed the m3 shorts when routing over cap arrays.
>
> Still many opens and shorts in wiring up the cap arrays. We probably need
> a meeting to figure out what is going on.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ALIGN-2Danaloglayout_ALIGN-2Dpublic_pull_215-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAKIMLZWWF5HJXLZP2EZR4J3QPW4SHA5CNFSM4JCTHU52YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCIT7GCY-23pullrequestreview-2D304608011&d=DwMCaQ&c=u6LDEWzohnDQ01ySGnxMzg&r=ZMUCBc0sWFUe5MzbQ3jL6w&m=vU-wdn6Vh65ABlMRDgem-9J-iN5k-kPT1A3kZrCfNKI&s=x1K2vRF1jaA5zkPwBWKdDMTg4ULDg-UB1zaq4LZRif8&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AKIMLZW6INWADOT3VTQJHQLQPW4SHANCNFSM4JCTHU5Q&d=DwMCaQ&c=u6LDEWzohnDQ01ySGnxMzg&r=ZMUCBc0sWFUe5MzbQ3jL6w&m=vU-wdn6Vh65ABlMRDgem-9J-iN5k-kPT1A3kZrCfNKI&s=dbVJax8jCo7IHZ2GPF5n1HyPOguXZUIrMqzUrxq7GaQ&e=>
> .
>

Lastdayends commented:
Hi Steve,

I will give some figures to explain the details about 'By the way, the
location of the via generated by the layout visualization tool is not
correct.'.

Yaguang

On Mon, Oct 21, 2019 at 10:11 AM Steve Burns <notifications@github.com>
wrote:

> can you tell me more about this?
> "By the way, the location of the via generated by the layout visualization
> tool is not correct."
> We need to get this right.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ALIGN-2Danaloglayout_ALIGN-2Dpublic_pull_215-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAKIMLZQB2YH3KF5F2FVOHRDQPXBAPA5CNFSM4JCTHU52YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEB2VJWI-23issuecomment-2D544560345&d=DwMCaQ&c=u6LDEWzohnDQ01ySGnxMzg&r=ZMUCBc0sWFUe5MzbQ3jL6w&m=_KAsIQluqD5zxv2WQR20RnwwqtXVWtDJdiwfs3N6NHU&s=k3F-honkGr-rB3wfBHD82T1NgGvERXpfVBjK33Z7l24&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AKIMLZUKBC3OF7VZOGQ4S6DQPXBAPANCNFSM4JCTHU5Q&d=DwMCaQ&c=u6LDEWzohnDQ01ySGnxMzg&r=ZMUCBc0sWFUe5MzbQ3jL6w&m=_KAsIQluqD5zxv2WQR20RnwwqtXVWtDJdiwfs3N6NHU&s=Ox5BnSBfJ346V5YF-z_yvLgXSFsNKQL5Nd0157pEOZU&e=>
> .
>

stevenmburns commented:
What is an intervia? I thought is was an obstruction. Does it have a netname attached to it?
What code should I put in `WriteViewerJSON` (in cap_placer/capplacer.cpp) to write out all these vias? I was following the code in `WriteJSON` (which writes out the GDS json.)

Lastdayends commented:
Hi Steve,

It does not a netname attached to the vias.

You can click the option when visualizing the layout. Here I give an
example in switched capacitor combination.

[image: image.png][image: image.png] [image: image.png]
                  a. net 72                                    b. without
!Intervias             c. with !Intervias

Figure a is the origin visualization of net 72. b and c are locally
visualization of a in two case, without !Intervias and with !Intervias. In
b, you may see some opens and shorts due to lack of vias, but in c not. (
!Intervias is an option in the webpage....)

Is this the open and short problem that you mentioned?

Yaguang

stevenmburns commented:
Those vias aren’t in the Viewer JSON. I want them to be. How should I change the C++ code that produces the Viewer JSON to get these? Where are they stored?

From: Yaguang Li [mailto:notifications@github.com]
Sent: Monday, October 21, 2019 12:56 PM
To: ALIGN-analoglayout/ALIGN-public <ALIGN-public@noreply.github.com>
Cc: Burns, Steven M <steven.m.burns@intel.com>; Mention <mention@noreply.github.com>
Subject: Re: [ALIGN-analoglayout/ALIGN-public] correct unit cap (cap 12) pin content (#215)

Hi Steve,

It does not a netname attached to the vias.

You can click the option when visualizing the layout. Here I give an
example in switched capacitor combination.

[image: image.png][image: image.png] [image: image.png]
a. net 72 b. without
!Intervias c. with !Intervias

Figure a is the origin visualization of net 72. b and c are locally
visualization of a in two case, without !Intervias and with !Intervias. In
b, you may see some opens and shorts due to lack of vias, but in c not. (
!Intervias is an option in the webpage....)

Is this the open and short problem that you mentioned?

Yaguang


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ALIGN-analoglayout/ALIGN-public/pull/215?email_source=notifications&email_token=AALRI7T6NG75A3KEOMEYUXTQPYCO3A5CNFSM4JCTHU52YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEB3SXBY#issuecomment-544680839>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AALRI7RQ5SRU2KFYWIS5G6TQPYCO3ANCNFSM4JCTHU5Q>.

Lastdayends commented:
Hi Steve,

The problem, which I mentioned, 'By the way, the location of the via
generated by the layout visualization tool is not correct.', seems to
relate to placer. The figure a is an mirrored figure compared to what I
expected. So that's why I said the layout visualization tool is not
correct. But now I guess the placer make a mirror of the cap. Anyway, if no
open and short problem, you can ignore the problem it.

[image: image.png]

                  a. layout visualization

Yaguang

On Mon, Oct 21, 2019 at 10:45 AM Yaguang Li <liyg@tamu.edu> wrote:

> Hi Steve,
>
> I will give some figures to explain the details about 'By the way, the
> location of the via generated by the layout visualization tool is not
> correct.'.
>
> Yaguang
>
> On Mon, Oct 21, 2019 at 10:11 AM Steve Burns <notifications@github.com>
> wrote:
>
>> can you tell me more about this?
>> "By the way, the location of the via generated by the layout
>> visualization tool is not correct."
>> We need to get this right.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ALIGN-2Danaloglayout_ALIGN-2Dpublic_pull_215-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAKIMLZQB2YH3KF5F2FVOHRDQPXBAPA5CNFSM4JCTHU52YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEB2VJWI-23issuecomment-2D544560345&d=DwMCaQ&c=u6LDEWzohnDQ01ySGnxMzg&r=ZMUCBc0sWFUe5MzbQ3jL6w&m=_KAsIQluqD5zxv2WQR20RnwwqtXVWtDJdiwfs3N6NHU&s=k3F-honkGr-rB3wfBHD82T1NgGvERXpfVBjK33Z7l24&e=>,
>> or unsubscribe
>> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AKIMLZUKBC3OF7VZOGQ4S6DQPXBAPANCNFSM4JCTHU5Q&d=DwMCaQ&c=u6LDEWzohnDQ01ySGnxMzg&r=ZMUCBc0sWFUe5MzbQ3jL6w&m=_KAsIQluqD5zxv2WQR20RnwwqtXVWtDJdiwfs3N6NHU&s=Ox5BnSBfJ346V5YF-z_yvLgXSFsNKQL5Nd0157pEOZU&e=>
>> .
>>
>

Lastdayends commented:
Hi Steve,

In the code, I think you have already written the vias out. You can check
the code: capplacer.cpp from line 1882 to 1920 in the
function WriteViewerJSON(). I have checked the function from line 1882 to
1920, logically they are correct.

Yaguang

On Mon, Oct 21, 2019 at 3:00 PM Steve Burns <notifications@github.com>
wrote:

> Those vias aren’t in the Viewer JSON. I want them to be. How should I
> change the C++ code that produces the Viewer JSON to get these? Where are
> they stored?
>
> From: Yaguang Li [mailto:notifications@github.com]
> Sent: Monday, October 21, 2019 12:56 PM
> To: ALIGN-analoglayout/ALIGN-public <ALIGN-public@noreply.github.com>
> Cc: Burns, Steven M <steven.m.burns@intel.com>; Mention <
> mention@noreply.github.com>
> Subject: Re: [ALIGN-analoglayout/ALIGN-public] correct unit cap (cap 12)
> pin content (#215)
>
> Hi Steve,
>
> It does not a netname attached to the vias.
>
> You can click the option when visualizing the layout. Here I give an
> example in switched capacitor combination.
>
> [image: image.png][image: image.png] [image: image.png]
> a. net 72 b. without
> !Intervias c. with !Intervias
>
> Figure a is the origin visualization of net 72. b and c are locally
> visualization of a in two case, without !Intervias and with !Intervias. In
> b, you may see some opens and shorts due to lack of vias, but in c not. (
> !Intervias is an option in the webpage....)
>
> Is this the open and short problem that you mentioned?
>
> Yaguang
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub<
> https://github.com/ALIGN-analoglayout/ALIGN-public/pull/215?email_source=notifications&email_token=AALRI7T6NG75A3KEOMEYUXTQPYCO3A5CNFSM4JCTHU52YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEB3SXBY#issuecomment-544680839>,
> or unsubscribe<
> https://github.com/notifications/unsubscribe-auth/AALRI7RQ5SRU2KFYWIS5G6TQPYCO3ANCNFSM4JCTHU5Q>.
>
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ALIGN-2Danaloglayout_ALIGN-2Dpublic_pull_215-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAKIMLZS2U65YXBGYC3DZPFDQPYC47A5CNFSM4JCTHU52YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEB3TC4Q-23issuecomment-2D544682354&d=DwMFaQ&c=u6LDEWzohnDQ01ySGnxMzg&r=ZMUCBc0sWFUe5MzbQ3jL6w&m=3FbrPugp98MR2NlBjGGPhIlOV_9_d863ZOwYBk3iXlo&s=FB_PvzCN2pA1BBmUNVcnArx0sZLBR0sRHDQZ2dVxKec&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AKIMLZWZOZ236NWUMZI6Q6TQPYC47ANCNFSM4JCTHU5Q&d=DwMFaQ&c=u6LDEWzohnDQ01ySGnxMzg&r=ZMUCBc0sWFUe5MzbQ3jL6w&m=3FbrPugp98MR2NlBjGGPhIlOV_9_d863ZOwYBk3iXlo&s=fNsbMGTGbzhQKISne2N3y-kvZFw16oYhgYOmlt5HehM&e=>
> .
>

stevenmburns commented:
This is the code I added to put vias in the viewer JSON. If it is logically correct, then why aren't there vias in this JSON?

Lastdayends commented:
The via name needs a transformation, which is done in GDSwriter while not
in JsonViewer. I have fixed the bug. Now the vias should be shown.

Yaguang

On Mon, Oct 21, 2019 at 4:21 PM Steve Burns <notifications@github.com>
wrote:

> This is the code I added to put vias in the viewer JSON. If it is
> logically correct, then why aren't there vias in this JSON?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ALIGN-2Danaloglayout_ALIGN-2Dpublic_pull_215-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAKIMLZW4DBUHH2QSRTPEF3TQPYMOPA5CNFSM4JCTHU52YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEB32UMI-23issuecomment-2D544713265&d=DwMCaQ&c=u6LDEWzohnDQ01ySGnxMzg&r=ZMUCBc0sWFUe5MzbQ3jL6w&m=NVFRmsuNAHzdM3eoLrnXBPoVphuQHOvE83kXler7Ufw&s=cJZjP1kahouCI_zzNZqg0UzAKxBOcFutEvwZ7c7vOFo&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AKIMLZR27BEQRLJSB3LWAD3QPYMOPANCNFSM4JCTHU5Q&d=DwMCaQ&c=u6LDEWzohnDQ01ySGnxMzg&r=ZMUCBc0sWFUe5MzbQ3jL6w&m=NVFRmsuNAHzdM3eoLrnXBPoVphuQHOvE83kXler7Ufw&s=yHCBxVzgldRuyqNkHWsmo9zI95Qg96bKDuaa6-iYdrw&e=>
> .
>
